File: api/download.mjs
import { spawn } from "node:child_process";
import { createReadStream } from "node:fs";
import { mkdtemp, readdir, rm } from "node:fs/promises";
import { tmpdir } from "node:os";
import { join } from "node:path";

export const config = { maxDuration: 300 };

const binary = join(process.cwd(), "node_modules", "youtube-dl-exec", "bin", "yt-dlp");
const allowedKinds = new Set(["video", "audio"]);
const quality = new Set(["1080", "720", "480"]);

function run(binaryPath, args) {
  return new Promise((resolve, reject) => {
    const child = spawn(binaryPath, args, { windowsHide: true });
    let error = "";
    child.stderr.on("data", (chunk) => { error += chunk; });
    child.on("error", reject);
    child.on("close", (code) => code === 0 ? resolve() : reject(new Error(error || "yt-dlp could not download this link.")));
  });
}

export default async function handler(request, response) {
  if (request.method !== "POST") return response.status(405).json({ error: "POST required." });
  const { url, kind = "video", quality: selectedQuality = "1080" } = request.body || {};
  let parsed;
  try { parsed = new URL(url); } catch { return response.status(400).json({ error: "Enter a valid media URL." }); }
  if (!/^https?:$/.test(parsed.protocol) || !allowedKinds.has(kind)) return response.status(400).json({ error: "Unsupported request." });

  const height = quality.has(selectedQuality) ? selectedQuality : "1080";
  const directory = await mkdtemp(join(tmpdir(), "clipkit-"));
  const output = join(directory, "clipkit.%(ext)s");
  const format = kind === "audio"
    ? "bestaudio[ext=m4a]/bestaudio"
    : `best[ext=mp4][height<=${height}]/best[height<=${height}]`;
  try {
    await run(binary, [parsed.toString(), "--no-playlist", "--no-progress", "--restrict-filenames", "-f", format, "-o", output]);
    const [file] = (await readdir(directory)).filter((name) => name.startsWith("clipkit."));
    if (!file) throw new Error("No file was produced.");
    const extension = file.split(".").pop() || "bin";
    response.setHeader("Content-Type", extension === "mp4" ? "video/mp4" : extension === "m4a" ? "audio/mp4" : "application/octet-stream");
    response.setHeader("Content-Disposition", `attachment; filename="clipkit.${extension}"`);
    const stream = createReadStream(join(directory, file));
    stream.on("error", () => response.destroy());
    response.on("finish", () => rm(directory, { recursive: true, force: true }));
    return stream.pipe(response);
  } catch (error) {
    await rm(directory, { recursive: true, force: true });
    return response.status(422).json({ error: "Download failed. Confirm the link is publicly available and try again." });
  }
}

File: api/download.py
import json
import os
import shutil
import tempfile
from http.server import BaseHTTPRequestHandler
from urllib.parse import urlparse

import yt_dlp


class handler(BaseHTTPRequestHandler):
    def do_POST(self):
        try:
            size = int(self.headers.get("Content-Length", "0"))
            payload = json.loads(self.rfile.read(size).decode("utf-8"))
            media_url = payload.get("url", "")
            kind = payload.get("kind", "video")
            selected_quality = payload.get("quality", "1080")
            parsed = urlparse(media_url)
            if parsed.scheme not in {"http", "https"} or not parsed.netloc or kind not in {"video", "audio"}:
                return self.respond_json(400, {"error": "Enter a valid media URL and format."})

            height = selected_quality if selected_quality in {"1080", "720", "480"} else "1080"
            directory = tempfile.mkdtemp(prefix="clipkit-")
            try:
                options = {
                    "outtmpl": os.path.join(directory, "clipkit.%(ext)s"),
                    "noplaylist": True,
                    "quiet": True,
                    "no_warnings": True,
                    "restrictfilenames": True,
                    "format": "bestaudio[ext=m4a]/bestaudio" if kind == "audio" else f"best[ext=mp4][height<={height}]/best[height<={height}]",
                }
                with yt_dlp.YoutubeDL(options) as downloader:
                    downloader.download([media_url])
                files = [name for name in os.listdir(directory) if name.startswith("clipkit.")]
                if not files:
                    raise RuntimeError("No file was produced.")
                filename = files[0]
                path = os.path.join(directory, filename)
                content_type = "video/mp4" if filename.endswith(".mp4") else "audio/mp4" if filename.endswith(".m4a") else "application/octet-stream"
                self.send_response(200)
                self.send_header("Content-Type", content_type)
                self.send_header("Content-Disposition", f'attachment; filename="{filename}"')
                self.send_header("Content-Length", str(os.path.getsize(path)))
                self.end_headers()
                with open(path, "rb") as file:
                    shutil.copyfileobj(file, self.wfile)
            finally:
                shutil.rmtree(directory, ignore_errors=True)
        except Exception as error:
            print(f"[clipkit:download-failed] {type(error).__name__}: {str(error)[:2000]}")
            self.respond_json(422, {"error": "Download failed. Confirm the link is publicly available and try again."})

    def respond_json(self, status, body):
        encoded = json.dumps(body).encode("utf-8")
        self.send_response(status)
        self.send_header("Content-Type", "application/json; charset=utf-8")
        self.send_header("Content-Length", str(len(encoded)))
        self.end_headers()
        self.wfile.write(encoded)
